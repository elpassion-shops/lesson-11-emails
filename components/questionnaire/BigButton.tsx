
export function BigButton() {
    return (
      <button
        type="submit"
        className={"confirm-button bg-rose-600 px-4 py-2 rounded-md text-white drop-shadow-md"}
      >
        Wy≈õlij odpowiedzi
      </button>
    );
  }
  