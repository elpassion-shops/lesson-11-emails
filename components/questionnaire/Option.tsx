import { useState } from "react";

interface OptionProps {
  answear: number;
  className: string;
  onClick(): void;
}

export function Option(props: OptionProps) {
  return (
    <button
      className={props.className}
      onClick={() => {
        props.onClick();
      }}
    >
      {props.answear}
    </button>
  );
}
