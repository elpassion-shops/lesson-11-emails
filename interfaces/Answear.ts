export interface Answear {
  rating: number;
  key: number;
}
