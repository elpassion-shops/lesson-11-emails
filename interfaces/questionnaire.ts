import {
  IsEmail,
  IsInt,
  IsNotIn,
  IsOptional,
  IsString,
  Max,
  Min,
  ValidateNested,
} from "class-validator";

export interface IAnswer {
  isOpen: boolean;
  value?: number | string | null;
  questionId?: number;

}

export interface IQuestion {
  id: number;
  question: string;
  answer?: IAnswer;
  isOpen?: boolean | null;
}

export interface IQuestionnaire {
  id: number;
  title: string;
  questions: IQuestion[];
  email?: string | null;
}

export interface IQuestionnaireResponse {
  questionnaire: IQuestionnaire;
}

export interface IQuestionnaireRequest {
  id: number;
}

export interface IAnswersRequest {
  answers: IAnswer[];
  email: string;
}

export class Answer implements IAnswer{
  isOpen: boolean;
  questionId: number;
  @IsNotIn([0,"",null,undefined])
  value: string | number;

}

export class AnswerRequest {
  @ValidateNested({each: true})
  answers: Answer[];

  @IsEmail()
  email: string;
}

export class AnswerClose implements IAnswer {
  isOpen = false;

  @IsInt()
  @Min(1)
  @Max(5)
  value?: number;

  constructor(value: number) {
    this.value = value;
  }
}

export class AnswerOpen implements IAnswer {
  isOpen = true;

  @IsString()
  value?: string;

  constructor(value: string) {
    this.value = value;
  }
}

export class Question implements IQuestion {
  @IsInt()
  @Min(1)
  id: number;

  @IsString()
  question: string;

  @ValidateNested()
  answer: IAnswer;

  constructor(id: number, question: string, answer: IAnswer) {
    this.id = id;
    this.question = question;
    this.answer = answer;
  }
}

export class Questionnaire implements IQuestionnaire {
  @IsInt()
  @Min(1)
  id;

  @IsString()
  title;

  @ValidateNested({ each: true })
  questions;

  @IsOptional()
  @IsEmail()
  email?;

  constructor(
    id: number,
    title: string,
    questions: IQuestion[],
    email?: string | null
  ) {
    this.id = id;
    this.title = title;
    this.questions = questions;
    this.email = email;
  }
}
