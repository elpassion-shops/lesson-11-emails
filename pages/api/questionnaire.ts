import "reflect-metadata";
import { validate, validateOrReject } from "class-validator";
import type { NextApiRequest, NextApiResponse } from "next";
import { Request } from "../../dtos/request";
import { plainToClass } from "class-transformer";

export default async function handler(
  req: NextApiRequest,
  res: NextApiResponse
) {
  try {
    const data = req.body;
    const request = plainToClass(Request, data);
    await validateOrReject(request);
    console.log(request);
    res.status(201).json({});
  } catch (error) {
    res.status(400).json(error);
  }
}
