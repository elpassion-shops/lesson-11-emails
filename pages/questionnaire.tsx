import { useRouter } from "next/router";
import React from "react";
import ReactDOM from "react-dom";
import { useForm, SubmitHandler } from "react-hook-form";

enum GenderEnum {
    female = "female",
    male = "male",
    other = "other"
  }
  interface IFormInput {
    firstName: String;
    gender: GenderEnum;
  }
  

//const resolver = classValidatorResolver
function Questionnaire() {

  const { register, handleSubmit } = useForm<IFormInput>();
  const onSubmit = (data: IFormInput) => console.log(data);

  return (
    <form >
      <label>First Name</label>
      <input {...register("firstName")} />
      <label>Gender Selection</label>
      <select {...register("gender")}>
        <option value="female">female</option>
        <option value="male">male</option>
        <option value="other">other</option>
      </select>
      <input type="submit" />
    </form>
  );
}

  export default Questionnaire;
  