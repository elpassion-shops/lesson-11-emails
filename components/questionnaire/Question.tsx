import { useEffect, useState } from "react";
import { Answear } from "../../interfaces/Answear";
import { Option } from "./Option";

interface QuestionProps {
  text: string;
  hint: string;
  answears: Answear[];
  selected: number;
  onClick(answear: number): void;
}

export function Question(props: QuestionProps) {
  return (
    <div className="question">
      <p>{props.text}</p>
      <p className="hint">{props.hint}</p>
      <div className="options">
        {props.answears.map((a) => {
          return (
            <Option
              className={props.selected === a.rating? "selected" : "option-button"}
              onClick={() => {
                props.onClick(a.rating);
              }}
              answear={a.rating}
              key={a.key}
            />
          );
        })}
      </div>
    </div>
  );
}
