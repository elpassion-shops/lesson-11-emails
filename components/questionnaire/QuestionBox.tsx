import { useState } from "react";
import { Answear } from "../../interfaces/Answear";
import { Question } from "./Question";

export function QuestionBox() {
  interface QuestionWithAnswear {
    question: string;
    answear: number;
  }

  interface Question {
    question: string;
    hint: string;
    key: number;
    selected: number;
    answears: Answear[];
  }

  const [questions, setQuestions] = useState<Question[]>([
    {
      question:
        "Jak podobał Ci się nasz staż? (ostatnia szansa na zmianę zdania! :D)",
      hint: "(1 - w ogóle, 2 - niezbyt, 3 - średnio, 4 - był ok, 5 - bardzo)",
      key: Math.random(),
      selected: 0,
      answears: [
        { rating: 1, key: Math.random()},
        { rating: 2, key: Math.random()},
        { rating: 3, key: Math.random()},
        { rating: 4, key: Math.random()},
        { rating: 5, key: Math.random()}
      ],
    },
    {
      question: "Jak dobrze był Twoim zdaniem przygotowany prowadzący?",
      hint: "(1 - fatalnie, 2 - kiepsko, 3 - średnio, 4 - był ok, 5 - świetnie",
      key: Math.random(),
      selected: 0,
      answears: [
        { rating: 1, key: Math.random()},
        { rating: 2, key: Math.random()},
        { rating: 3, key: Math.random()},
        { rating: 4, key: Math.random()},
        { rating: 5, key: Math.random()},
      ],
    },
    {
      question: "Czy uważasz, że zdobyta wiedza przyda Ci się w życiu?",
      hint: "(1 - w ogóle mi się nie przyda, 2 - raczej nie, 3 - może, 4 - myślę, że tak, 5 - na pewno!",
      key: Math.random(),
      selected: 0,
      answears: [
        { rating: 1, key: Math.random()},
        { rating: 2, key: Math.random()},
        { rating: 3, key: Math.random()},
        { rating: 4, key: Math.random()},
        { rating: 5, key: Math.random()}
      ],
    },
  ]);

  const [questionsWithAnswears, setQuestionsWithAnswers] = useState<
    QuestionWithAnswear[]
  >([]);

  function handleClick(question: string, answear: number) {
    setQuestionsWithAnswers((prevState) =>
      prevState.filter((q) => q.question !== question)
    );
    setQuestionsWithAnswers((prevState) =>
      prevState.concat({ question: question, answear: answear })
    );

    setQuestions((prevState) => {
      return prevState.map((q) => {
        if (q.question !== question) return q;
        return {question: q.question, hint: q.hint, key: q.key, selected: answear, answears: q.answears}
        })
  })
}
  return (
    <div className="question-box">
      {questions.map((q) => {
        return (
          <Question
            onClick={(answear) => handleClick(q.question, answear)}
            text={q.question}
            hint={q.hint}
            key={q.key}
            answears={q.answears}
            selected={q.selected}
          />
        );
      })}
    </div>
  );
}
