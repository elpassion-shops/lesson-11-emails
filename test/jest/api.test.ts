import { validate } from "class-validator";
import { Answer, AnswerRequest } from "../../interfaces/questionnaire";

function generateAnswerRequest(overrides: { email?: any; answers?: any }) {
  const request = new AnswerRequest();
  request.email =
    overrides.email === undefined ? "bugajewska.m@gmail.com" : overrides.email;
  request.answers = overrides.answers === undefined ? [] : overrides.answers;
  return request;
}

function generateAnswer(overrides: { questionId?: any; value?: any }) {
  const answer = new Answer();
  answer.questionId =
    overrides.questionId === undefined ? 1 : overrides.questionId;
  answer.value = overrides.value === undefined ? [] : overrides.value;
  return answer;
}

test("passes with correct request", async () => {
  const request = generateAnswerRequest({});
  expect(await validate(request)).toEqual([]);
});

test("passes with correct answers", async () => {
  const request = generateAnswerRequest({
    answers: [generateAnswer({ value: 1, questionId: 1 })],
  });
  expect(await validate(request)).toMatchSnapshot();
});

test("fails without email", async () => {
  const request = generateAnswerRequest({ email: null });
  expect(await validate(request)).toEqual([
    expect.objectContaining({
      constraints: {
        isEmail: "email must be an email",
      },
      property: "email",
    }),
  ]);
});

test("fails with incorrect email", async () => {
  const request = generateAnswerRequest({ email: "bugajewskam.com" });
  expect(await validate(request)).toEqual([
    expect.objectContaining({
      constraints: {
        isEmail: "email must be an email",
      },
      property: "email",
    }),
  ]);
});

test("fails with empty answers", async () => {
  const request = generateAnswerRequest({ answers: null });
  expect(await validate(request)).toEqual([
    expect.objectContaining({
      children: [
        expect.objectContaining({
          constraints: {
            nestedValidation:
              "each value in nested property answers must be either object or array",
          },
          property: "answers",
        }),
      ],
    }),
  ]);
});

test("fails with incorrect answer", async () => {
  const request = generateAnswerRequest({ answers: [new Answer()] });
  expect(await validate(request)).toMatchSnapshot();
});

test("passes with incorrect answer value", async () => {
  const request = generateAnswerRequest({
    answers: [generateAnswer({ value: 0 })],
  });
  expect(await validate(request)).toMatchSnapshot();
});
