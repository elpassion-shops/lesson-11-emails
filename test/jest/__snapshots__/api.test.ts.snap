// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fails with incorrect answer 1`] = `
Array [
  ValidationError {
    "children": Array [
      ValidationError {
        "children": Array [
          ValidationError {
            "children": Array [],
            "constraints": Object {
              "isNotIn": "value should not be one of the following values: 0, 0, , , ",
            },
            "property": "value",
            "target": Answer {},
            "value": undefined,
          },
        ],
        "property": "0",
        "target": Array [
          Answer {},
        ],
        "value": Answer {},
      },
    ],
    "property": "answers",
    "target": AnswerRequest {
      "answers": Array [
        Answer {},
      ],
      "email": "bugajewska.m@gmail.com",
    },
    "value": Array [
      Answer {},
    ],
  },
]
`;

exports[`fails with incorrect email 1`] = `
Array [
  ValidationError {
    "children": Array [],
    "constraints": Object {
      "isEmail": "email must be an email",
    },
    "property": "email",
    "target": AnswerRequest {
      "answers": Array [],
      "email": "bugajewskam.com",
    },
    "value": "bugajewskam.com",
  },
]
`;

exports[`passes with correct answers 1`] = `Array []`;

exports[`passes with incorrect answer value 1`] = `
Array [
  ValidationError {
    "children": Array [
      ValidationError {
        "children": Array [
          ValidationError {
            "children": Array [],
            "constraints": Object {
              "isNotIn": "value should not be one of the following values: 0, 0, , , ",
            },
            "property": "value",
            "target": Answer {
              "questionId": 1,
              "value": 0,
            },
            "value": 0,
          },
        ],
        "property": "0",
        "target": Array [
          Answer {
            "questionId": 1,
            "value": 0,
          },
        ],
        "value": Answer {
          "questionId": 1,
          "value": 0,
        },
      },
    ],
    "property": "answers",
    "target": AnswerRequest {
      "answers": Array [
        Answer {
          "questionId": 1,
          "value": 0,
        },
      ],
      "email": "bugajewska.m@gmail.com",
    },
    "value": Array [
      Answer {
        "questionId": 1,
        "value": 0,
      },
    ],
  },
]
`;
